File: tugashesti.java
public class tugashesti {
    public static void main(String[] args) throws java.io.IOException {
        // Kapasitas 10 baris, 2 kolom (NIM & Nama)
        String[][] data = new String[10][2];
        int count = 0; //
        int capacity = 10;
        int pilihan = 0;

        // ================= HEADER IDENTITAS =================
        System.out.println("\n=================================================");
        System.out.println("            TUGAS STRUKTUR DATA ARRAY            ");
        System.out.println("=================================================");
        System.out.println(" Nama    : Hesti Febriyani");
        System.out.println(" Kelas   : Informatika A");
        System.out.println(" NIM     : 24106050010"); 
        System.out.println(" Matkul  : Struktur Data");
        System.out.println("=================================================");
        System.out.println(" >> Tekan ENTER untuk masuk ke Menu Program...");
        
        while (System.in.read() != '\n');

        // Loop berjalan selama pilihan bukan 9
        while (pilihan != 9) {
            System.out.println("\n=== MENU MAHASISWA (Count: " + count + "/" + capacity + ") ===");
            System.out.println("1. Insert at beginning    4. Delete from beginning");
            System.out.println("2. Insert at position     5. Delete given position");
            System.out.println("3. Insert at end          6. Delete from end");
            System.out.println("7. Delete first (NIM)     8. Show data");
            System.out.println("9. Exit");
            System.out.print("Pilih menu: ");

            int inputChar = System.in.read();
            pilihan = inputChar - '0';
            while (System.in.read() != '\n'); 

            switch (pilihan) {
                case 1: case 2: case 3:
                    if (count < capacity) {
                        int targetPos = (pilihan == 1) ? 0 : (pilihan == 3) ? count : -1;
                        if (pilihan == 2) {
                            System.out.print("Masukkan posisi index (0-" + count + "): ");
                            int pChar = System.in.read();
                            targetPos = pChar - '0';
                            while (System.in.read() != '\n');
                        }
                        if (targetPos >= 0 && targetPos <= count) {
                            System.out.print("Masukkan NIM : ");
                            String nim = ""; int c;
                            while ((c = System.in.read()) != '\n' && c != '\r') { nim += (char) c; }
                            if (c == '\r') System.in.read();

                            System.out.print("Masukkan Nama: ");
                            String nama = "";
                            while ((c = System.in.read()) != '\n' && c != '\r') { nama += (char) c; }
                            if (c == '\r') System.in.read();

                            for (int i = count; i > targetPos; i--) {
                                data[i][0] = data[i-1][0];
                                data[i][1] = data[i-1][1];
                            }
                            data[targetPos][0] = nim;
                            data[targetPos][1] = nama;
                            count++; //
                            System.out.println("Data berhasil ditambahkan.");
                        }
                    } else { System.out.println("!! Array Penuh !!"); }
                    break;

                case 4: // Delete from beginning
                    if (count > 0) {
                        for (int i = 0; i < count - 1; i++) {
                            data[i][0] = data[i+1][0];
                            data[i][1] = data[i+1][1];
                        }
                        count--; //
                        System.out.println("Data index 0 berhasil dihapus.");
                    }
                    break;

                case 5: // Delete given position
                    if (count > 0) {
                        System.out.print("Masukkan index yang dihapus: ");
                        int dPos = System.in.read() - '0';
                        while (System.in.read() != '\n');
                        if (dPos >= 0 && dPos < count) {
                            for (int i = dPos; i < count - 1; i++) {
                                data[i][0] = data[i+1][0];
                                data[i][1] = data[i+1][1];
                            }
                            count--; //
                        }
                    }
                    break;

                case 6: // Delete from end
                    if (count > 0) {
                        count--; //
                        System.out.println("Data terakhir berhasil dihapus.");
                    }
                    break;

                case 7: // Delete first occurrence (NIM)
                    if (count > 0) {
                        System.out.print("Masukkan NIM yang ingin dihapus: ");
                        String sNim = ""; int baca;
                        while ((baca = System.in.read()) != '\n' && baca != '\r') { sNim += (char) baca; }
                        if (baca == '\r') System.in.read();
                        int idxKetemu = -1;
                        for (int i = 0; i < count; i++) {
                            if (data[i][0] != null && data[i][0].equals(sNim)) {
                                idxKetemu = i; break;
                            }
                        }
                        if (idxKetemu != -1) {
                            for (int i = idxKetemu; i < count - 1; i++) {
                                data[i][0] = data[i+1][0];
                                data[i][1] = data[i+1][1];
                            }
                            count--; //
                            System.out.println("NIM " + sNim + " berhasil dihapus.");
                        }
                    }
                    break;

                case 8: // Show data
                    System.out.println("\n--- DAFTAR DATA MAHASISWA ---");
                    if (count == 0) System.out.println("Data masih kosong.");
                    for (int i = 0; i < count; i++) {
                        System.out.println("[" + i + "] NIM: " + data[i][0] + " | Nama: " + data[i][1]);
                    }
                    break;

                case 9: // Exit
                    System.out.println("Keluar dari program... Sampai jumpa Hesti!");
                    break;

                default:
                    System.out.println("!! Menu tidak valid !!");
                    break;
            }
        }
    }
}

File: tugashesti.py
# ================= HEADER IDENTITAS =================
print("\n=================================================")
print("            TUGAS STRUKTUR DATA ARRAY            ")
print("=================================================")
print(" Nama    : Hesti Febriyani")
print(" Kelas   : Informatika A")
print(" NIM     : 24106050010")
print(" Matkul  : Struktur Data")
print("=================================================")
input(" >> Tekan ENTER untuk masuk ke Menu Program...")

# Inisialisasi Array 2D (10 baris, 2 kolom) dengan nilai None
# Di Python, ini cara membuat fixed-size array manual
data = [[None for _ in range(2)] for _ in range(10)]
count = 0
capacity = 10
pilihan = 0

while pilihan != 9:
    print(f"\n=== MENU MAHASISWA (Count: {count}/{capacity}) ===")
    print("1. Insert at beginning    4. Delete from beginning")
    print("2. Insert at position     5. Delete given position")
    print("3. Insert at end          6. Delete from end")
    print("7. Delete first (NIM)     8. Show data")
    print("9. Exit")
    
    try:
        pilihan = int(input("Pilih menu: "))
    except ValueError:
        print("!! Masukkan angka 1-9 !!")
        continue

    if pilihan == 1 or pilihan == 2 or pilihan == 3:
        if count < capacity:
            # Tentukan posisi
            if pilihan == 1:
                target_pos = 0
            elif pilihan == 3:
                target_pos = count
            else:
                target_pos = int(input(f"Masukkan posisi index (0-{count}): "))

            if 0 <= target_pos <= count:
                nim = input("Masukkan NIM : ")
                nama = input("Masukkan Nama: ")

                # LOGIKA GESER KANAN (Shifting Manual)
                # Dimulai dari belakang ke depan
                for i in range(count, target_pos, -1):
                    data[i][0] = data[i-1][0]
                    data[i][1] = data[i-1][1]
                
                data[target_pos][0] = nim
                data[target_pos][1] = nama
                count += 1
                print("Data berhasil ditambahkan.")
            else:
                print("!! Posisi tidak valid !!")
        else:
            print("!! Array Penuh !!")

    elif pilihan == 4: # Delete beginning
        if count > 0:
            # LOGIKA GESER KIRI (Manual)
            for i in range(0, count - 1):
                data[i][0] = data[i+1][0]
                data[i][1] = data[i+1][1]
            count -= 1
            print("Data index 0 berhasil dihapus.")
        else:
            print("!! Data Kosong !!")

    elif pilihan == 5: # Delete given position
        if count > 0:
            d_pos = int(input("Masukkan index yang dihapus: "))
            if 0 <= d_pos < count:
                for i in range(d_pos, count - 1):
                    data[i][0] = data[i+1][0]
                    data[i][1] = data[i+1][1]
                count -= 1
                print(f"Data index {d_pos} berhasil dihapus.")
            else:
                print("!! Index tidak valid !!")

    elif pilihan == 6: # Delete end
        if count > 0:
            count -= 1
            print("Data terakhir berhasil dihapus.")
        else:
            print("!! Data Kosong !!")

    elif pilihan == 7: # Delete first occurrence (NIM)
        if count > 0:
            s_nim = input("Masukkan NIM yang ingin dihapus: ")
            idx_ketemu = -1
            # Cari manual
            for i in range(count):
                if data[i][0] == s_nim:
                    idx_ketemu = i
                    break
            
            if idx_ketemu != -1:
                for i in range(idx_ketemu, count - 1):
                    data[i][0] = data[i+1][0]
                    data[i][1] = data[i+1][1]
                count -= 1
                print(f"NIM {s_nim} berhasil dihapus.")
            else:
                print("!! NIM tidak ditemukan !!")

    elif pilihan == 8: # Show data
        print("\n--- DAFTAR DATA MAHASISWA ---")
        if count == 0:
            print("Data masih kosong.")
        for i in range(count):
            print(f"[{i}] NIM: {data[i][0]} | Nama: {data[i][1]}")

    elif pilihan == 9:
        print("Keluar dari program... Sampai jumpa Hesti!")

    else:
        print("!! Menu tidak valid !!")

print("Program Selesai.")